libfm/data/lxshortcut.1.in
.\" -*-nroff-*-
.TH LXSHORTCUT 1 "August 2013" "@PACKAGE@ @VERSION@" "lxshortcut manual"
.SH NAME
lxshortcut \- desktop entry file editor
.SH SYNOPSIS
.B lxshortcut
[ \-i
.I input-file
] [ \-o
.I output-file
]
.SH DESCRIPTION
lxshortcut is a simple application that uses LibFM file properties dialog
window to change properties of a desktop entry - icon, display name, path
and arguments of execution, etc.
.SH OPTIONS
At least one option should be specified and if \fIoutput-file\fP isn't
specified then \fIinput-file\fP will be used to make a result - either
full path or path in XDG specified user data directory for menu entries.
.TP 20
.BI \-i " input-file" "\fR,\fP \-\^\-input=" input-file
use \fIinput-file\fP as source for editing, it may be some XDG menu entry
if no full path is specified or custom path to some file
.TP
.BI \-o " output-file" "\fR,\fP \-\^\-output=" output-file
write result into \fIoutput-file\fP
.SH SEE ALSO
XDG Desktop Entry Specification
.SH AUTHOR
lxshortcut was written by Hong Jen Yee <pcman.tw@gmail.com> and modified
by Andriy Grytsenko <andrej@rep.kiev.ua>.
.PP
This manual page was written by Andriy Grytsenko <andrej@rep.kiev.ua>.

CHANGELOG

lxqt-build-tools-0.6.0 / 2018-01-25
===================================

  * Bump version to 0.6.0
  * Don't use automatic string conversions
  * Introduced lxqt-transupdate for translation updates
  * Improved cmake scripting
    - Set cmake_minimum_required to 3.1.0
    - Consider QT_SELECT - some distributions use this tool
    - Drop 0X and 11 support, require c++14
    - Remvoed QUIET from any finds
    - Cleaned up LXQtTranslateTs module
    - Removed translation pulling and cleanup external repos
      after translations moved back to their original repos
    - Remove link-time optimization for gcc since it breaks
      qobject_cast() in some cases.
    - Removed obsolete FindFm module

lxqt-build-tools-0.5.0 / 2018-05-21
===================================

  * Bump version to 0.5.0
  * Prevent in-source builds
  * Adds an PreventInSourceBuilds module
  * http --> https
  * Fixed some LXDE mentions
  * LXQtCompilerSettings: Set CMAKE_C_VISIBILITY_PRESET to hidden
  * Don't allow Qt foreach/Q_FOREACH
  * Update CMakeLists.txt
  * Require Qt 5.7.1
  * Adds a FindXKBCommon CMake module
  * Adds ECMFindModuleHelpers CMake module
  * Don't export github templates

0.4.0 / 2017-09-22
==================

  * Release 0.4.0: Update changelog
  * Bump version to 0.4.0
  * Added needed versions
  * remove version set from LXQtConfigVars.in
  * Move LXQt version to liblxqt
  * Fixes a FTBFS with superbuild mode
  * Adds a FindExif module
  * LXQtCompilerSettings: Fixes a typo in diagnostics-color
  * Adds (Ninja) colored diagnostics for GCC (>=4.9)
  * Enable strict iterators for debug builds

0.3.2 / 2017-01-14
==================

  * Release 0.3.2: Update changelog
  * Bump version to 0.3.2 - finally the build tools are arch: all (#15)
  * Warnings (#16)
  * Remove arched bits from version.cmake-file (#14)

0.3.1 / 2017-01-01
==================

  * Release 0.3.1: Update changelog
  * Bump LXQT_PATCH_VERSION and build tools version (#13)
  * 32 64 bit ness (#12)
  * Use "gcc-ar" and "gcc-ranlib" tools when building static libs with gcc link-time optimization. (#11)

0.3.0 / 2016-12-06
==================

  * Release 0.3.0: Update changelog
  * Adds an FindXCB CMake module
  * FindUDev: Add VERSION_STRING support
  * FindUDev: Set REQUIRED_VARS
  * FindUdev: Feed pkg_modules_modules results to find_library/path
  * FindUDev: Adds Feature Summary
  * Adds FindUDev module
  * Enables colored diagnostics for CLang/Ninja combination
  * Disable link-time optimization for clang++/llvm since after some testing it breaks QObject signal/slot sometimes. * Remove trailing spaces.
  * Use different link-time optimization compiler options for gcc and clang and move -Bsymbolic flags to linker flags.
  * Add more linker options to turn on linker-time optimizations for gcc & clang.
  * Set CMAKE_BUILD_TYPE to Release by default if it's value is not set.
  * FindGLIB: Improve gio-unix handling
  * Adds a FindGLIB CMake module
  * Adds a FindMenuCache CMake module
  * Adds a FindFm CMake module
  * move project name to top
  * Fixes the name of package config file name in superbuild mode

0.2.0 / 2016-10-30
==================

  * Release 0.2.0: Update changelog
  * Bump minor to 2 (#7)
  * Added Qt5Core to README.md (#6)
  * Adds an LXQtConfigVars module
  * Adds superbuild support
  * README.md: Update dependencies

0.1.0 / 2016-10-14
==================

  * Release 0.1.0: Update changelog
  * Adds LXQtCreatePortableHeaders module
  * Authors and license (#2)
  * Add README.md
  * Make lxqt-build-tools be usable

0.0.0 / 2016-09-30
==================

  * Remove cpack (#107)
  * Bump version ot 0.11.0 (#105)
  * Settings: Add "homemade" XDG_CONFIG_DIRS support
  * translations: Change default translations repository
  * Don't lock the screen when it's already locked.
  * LXQtTranslateTs.cmake: Enable PULL_TRANSLATIONS by default
  * LXQtTranslateTs.cmake: Use/update template in remote repo
  * .gitignore: Cleanup
  * build: Use external translations
  * translations: Fix external translations logic
  * translations: Add external translations support
  * Adds CMAKE_INSTALL_PREFIX to the LXQt Theme search paths
  * LXQtCompilerSettings: Add QT_USE_QSTRINGBUILDER to the compile definitions
  * TranslateDesktopTs: Adds UPDATE_OPTIONS options
  * TranslateDesktop: Remove lupdate absolute option
  * TranslateDesktop: Use -a switch in grep commands
  * Use CMAKE_INSTALL_FULL_DATAROOTDIR
  * Add LXQT_GRAPHICS_DIR directory definition
  * Adds FindXdgUserDirs CMake find_package module
  * Release 0.10.0
  * pluginTranslation: removed using QStringLiteral as it was causing SEGFAULT in static finalization time
  * pluginTranslation: add some symbol for referencing (for file to be not stripped out in static linking time)
  * Adds "automatic" plugin translations loader support
  * Rename LxQt to LXQt everywhere
  * Handles CMake policy CMP0063
  * Adds an lxqt_enable_target_exceptions() CMake function
  * Adds LXQtCompilerSettings CMake module
  * LXQtCreatePkgConfigFile.cmake: Makes install COMPONENT configurable
  * LXQtTranslateTs.cmake Adds install COMPONENT option
  * Adds Runtime install COMPONENT
  * Fixes an SuperBuild/NormalBuild FTBFS
  * CMake: Use CMAKE_AUTOUIC, drop qt5_wrap()
  * CMake: Quote variables
  * CMake: Removes unused variables
  * Removes LIB_INSTALL_DIR CACHE PATH variable
  * Use relative paths in install DESTINATION
  * Marks link libraries as PUBLIC
  * Use target_compile_definitions() instead of add_definitions()
  * Make Release the default build type
  * Simplifies build type matching
  * lxqt-config.cmake.in: Use find_dependency()
  * Updates the build system to the Targets infrastructure
  * Sort .desktop translations. Fix typos in the comments. Remove local dir artifact.
  * pkg-config: Adds Qt5Xdg to the required dependencies
  * Updates lxqt_create_pkgconfig_file()
  * Stops using Qt5Xdg dropped stuff
  * Adds LXQT_DATA_DIR configuration variable

libfm/src/tools/lxshortcut.c
/*
 *      lxshortcut.c
 *
 *      Copyright 2008  <pcman.tw@gmail.com>
 *      Copyright 2013 Andriy Grytsenko (LStranger) <andrej@rep.kiev.ua>
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; either version 2 of the License, or
 *      (at your option) any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License
 *      along with this program; if not, write to the Free Software
 *      Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *      MA 02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <fm-gtk.h>
#include <gtk/gtk.h>
#include <glib/gi18n.h>

#ifdef ENABLE_NLS
#include <libintl.h>
#endif

static gboolean no_display = FALSE;
static char* ifile = NULL;
static char* ofile = NULL;

static GOptionEntry opt_entries[] =
{
    {"no-display", 'n', 0, G_OPTION_ARG_NONE, &no_display, NULL, NULL},
    {"input", 'i', 0, G_OPTION_ARG_FILENAME, &ifile, N_("source file name or desktop id"), N_("SOURCE")},
    {"output", 'o', 0, G_OPTION_ARG_FILENAME, &ofile, N_("result file name"), N_("RESULT")},
    { NULL }
};

static gpointer dlg_init(GtkBuilder *ui, gpointer uidata, FmFileInfoList *files)
{
    gpointer widget;

    g_debug("dlg_init");
    /* FIXME: handle NoDisplay */
#define HIDE_WIDGET(x) widget = gtk_builder_get_object(ui, x); \
    if (widget) \
        gtk_widget_hide(widget)
    /* remove permissions tab */
    HIDE_WIDGET("permissions_tab");
    /* remove unwanted texts (times, target, etc.) */
    HIDE_WIDGET("target");
    HIDE_WIDGET("target_label");
    HIDE_WIDGET("total_size");
    HIDE_WIDGET("total_size_label");
    HIDE_WIDGET("size_on_disk");
    HIDE_WIDGET("size_on_disk_label");
    HIDE_WIDGET("mtime");
    HIDE_WIDGET("mtime_label");
    HIDE_WIDGET("atime");
    HIDE_WIDGET("atime_label");
    HIDE_WIDGET("ctime");
    HIDE_WIDGET("ctime_label");
#undef HIDE_WIDGET
    return (gpointer)-1;
}

static void dlg_finish(gpointer data, gboolean cancelled)
{
    /* FIXME: handle NoDisplay */
}

static FmFilePropertiesExtensionInit ext_table = {
    .init = &dlg_init,
    .finish = &dlg_finish
};

int main(int argc, char **argv)
{
    GError *err = NULL;
    FmConfig *config;
    GtkDialog *dlg;
    char *sfile = NULL, *tfile = NULL;
    const char * const *strv;
    char *bl[] = {"*", NULL};
    char *wl[] = {"gtk_file_prop:application/x-desktop", NULL};
    char *contents;
    gsize len;
    FmPath *path;
    FmFileInfo *fi;
    FmFileInfoList *files;

#ifdef ENABLE_NLS
    bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
    bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
    textdomain (GETTEXT_PACKAGE);
#endif

    if (argc == 1) {
        /* FIXME: we should just call 'lxshortcut --help' */
        g_print("Error: You need at least one option\n");
        g_print("for help run 'lxshortcut --help\n");
        return 1;
    }

    /* initialize GTK+ and parse the command line arguments */
    if (G_UNLIKELY(!gtk_init_with_args(&argc, &argv, " ", opt_entries, GETTEXT_PACKAGE, &err)))
    {
        g_print("Error: %s\n", err->message);
        return 1;
    }

    if (!ifile && !ofile)
        return 1;

    /* setup LibFM for its default config */
    config = fm_config_new();
    fm_config_load_from_file(config, "/dev/null");
    fm_gtk_init(config);
    /* set modules blacklist and allow only application/x-desktop */
    fm_config->modules_blacklist = g_strdupv(bl);
    fm_config->modules_whitelist = g_strdupv(wl);

    if (ofile)
        tfile = g_strdup(ofile);

    if (ifile)
    {
        // FIXME: this won't work for files without full path
        if (strchr(ifile, '/'))  /* if this is a file path */
        {
            if (!ofile) /* store to the same file if not otherwise specified. */
                tfile = g_strdup(ifile);
            else
                sfile = g_strdup(ifile);
        }
        else /* otherwise treat it as desktop id */
        {
            /* find in g_get_user_data_dir() */
            sfile = g_build_filename(g_get_user_data_dir(), "applications", ifile, NULL);
            if (!ofile)
                tfile = g_strdup(sfile);
            if (!g_file_test(sfile, G_FILE_TEST_IS_REGULAR))
            {
                /* otherwise find in g_get_system_data_dirs() */
                g_free(sfile);
                sfile = NULL;
                for (strv = g_get_system_data_dirs(); *strv; strv++)
                {
                    sfile = g_build_filename(*strv, "applications", ifile, NULL);
                    if (g_file_test(sfile, G_FILE_TEST_IS_REGULAR))
                        break;
                    g_free(sfile);
                    sfile = NULL;
                }
            }
        }
    }

    /* if we got sfile then copy found sfile into tfile */
    if (sfile && g_file_get_contents(sfile, &contents, &len, NULL))
    {
        g_file_set_contents(tfile, contents, len, NULL);
        /* FIXME: check errors */
        g_free(contents);
    }
    else if (!g_file_test(tfile, G_FILE_TEST_EXISTS))
        /* if no sfile is set then create an empty tfile */
        g_file_set_contents(tfile, "[Desktop Entry]\n"
                                   "Type=Application", -1, NULL);
    g_free(sfile);

    /* we have existing target file now with actual contents, let edit it */
    path = fm_path_new_for_path(tfile);
    /* get file info of it */
    fi = fm_file_info_new();
    fm_file_info_set_path(fi, path);
    fm_path_unref(path);
    fm_file_info_set_from_native_file(fi, tfile, &err);
    if (!fm_file_info_is_desktop_entry(fi))
    {
        g_print("Error: file %s is not a desktop entry file\n", tfile);
        g_free(tfile);
        fm_file_info_unref(fi);
        return 1;
    }
    g_free(tfile);
    files = fm_file_info_list_new();
    fm_file_info_list_push_tail_noref(files, fi);

    /* add own extension for it */
    fm_file_properties_add_for_mime_type("application/x-desktop", &ext_table);

    dlg = fm_file_properties_widget_new(files, TRUE);
    fm_file_info_list_unref(files);

    gtk_dialog_run(dlg);

    gtk_widget_destroy(GTK_WIDGET(dlg));

    fm_gtk_finalize();
    return 0;
}
